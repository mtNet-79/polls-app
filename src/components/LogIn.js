import { useState, useRef } from "react";
import { useNavigate } from "react-router-dom";
import { connect } from "react-redux";
// import { handleAddUser } from "../actions/users";
import { setAuthedUser } from "../actions/authedUser";
import { Link } from "react-router-dom";
// WHEN PAGE OPENS RENDER NEW USER OBJ WITH NEW RANDOM IMAGE
// IF USER SUBMITS NAME AND PASSWORD ADD TO THE OBJECT THEN SUBMIT THE OBJECT TO DISPATCH
const LogIn = (props) => {
  // const [user, setUser] = useState("");
  // const [password, setPassword] = useState("");
  const [formReady, setFormReady] = useState(false);
  const navigate = useNavigate();
  const { dispatch, image, users } = props;
  const userRef = useRef("");
  const passwordRef = useRef("");

  const checkForm = () => {
    
    if (
      userRef.current.value !== "" &&
      passwordRef.current.value !== "" &&
      passwordRef.current.value.length > 5
    )
      setFormReady(true);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (users[userRef.current.value]) {
      if (users[userRef.current.value].password === passwordRef.current.value) {
        // setUser(userRef.current.value);
        // setPassword(passwordRef.current.value);
        dispatch(setAuthedUser(userRef.current.value));
        navigate("/");
      } else {
        alert("Wrong password. Try again.");
      }
    } else {
      alert("This account does not exist. Add New User.");
    }

    // setUser("");
    // setPassword("");
  };
  // const toAddUserForm = () => {
  //   setFormReady(true)

  //   navigate("/add-user");
  // };
  //TODO: create random image

  return (
    <div className="container">
      <img src={image} alt="description" className="avatar" />
      <div className="add-new">
        <form onSubmit={handleSubmit} className="add-new">
          <label htmlFor="userName" className="form-label center">
            User
          </label>
          <input
            placeholder="User Name"
            type="text"
            name="userName"
            id="userName"
            // value={user}
            ref={userRef}
            onChange={checkForm}
          />
          <label htmlFor="password" className="form-label center">
            Password
          </label>
          <input
            placeholder="Password"
            type="text"
            name="password"
            id="password"
            // value={password}
            ref={passwordRef}
            onChange={checkForm}
          />
          <button
            type="submit"
            className="btn log-in-button"
            disabled={!formReady}
          >
            Log In
          </button>
        </form>

        <Link
          to="/add-user"
          className="btn add-new-btn"
          value="Add New User"
          // onClick={toAddUserForm}
        >
          Add New User
        </Link>
      </div>
    </div>
  );
};

const mapStateToProps = ({ authedUser, users }, { image }) => ({
  authedUser,
  image,
  users,
});
export default connect(mapStateToProps)(LogIn);
