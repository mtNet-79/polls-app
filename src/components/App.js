import { handleInitialData } from "../actions/shared";
import PollPage from "./PollPage";
import { useState, useEffect, Fragment } from "react";
import Dashboard from "./Dashboard";
import { connect } from "react-redux";
import Nav from './Nav';
import LogIn from "./LogIn";
import Leaderboard from "./Leaderboard";
import AddUser from "./AddUser";
import LoadingBar from "react-redux-loading-bar";
import { Routes, Route } from "react-router-dom";
import CreatePoll from "./CreatePoll";
import { useLocation, useNavigate, useParams } from "react-router-dom";

const withRouter = (Component) => {
  const ComponentWithRouterProp = (props) => {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  };

  return ComponentWithRouterProp;
};

const App = (props) => {
  console.log("app props: ", props);
  const [image, setImage] = useState("");
  useEffect(() => {
   
    const generateRandomAvatar = () => {
      fetch("https://avatar-endpoint.herokuapp.com/api/").then((res) => {
        res.blob().then((blobRes) => {
          const data = blobRes;
          const urlCreator = window.URL || window.webkitURL;
          const urlObj = urlCreator.createObjectURL(data);
          setImage(urlObj);
          console.log("trigged in generateavatar");
          props.dispatch(handleInitialData());
        });
      });
    }

    const alreadyLoggedIn = () => {
      console.log("triggerd at already")
      props.dispatch(handleInitialData());
    }
    
    props.authedUser === null 
      ? generateRandomAvatar() 
      : alreadyLoggedIn();
  }, []);



  return (
    <Fragment>
      <LoadingBar />       
        {props.loading ? null : (   
            <div className="container">              
              {props.location.pathname === "/add-user" 
                ? <AddUser image={image} />
                : props.authedUser
                  ? (
                    <Fragment>                   
                      <Nav />
                      <Routes>
                        <Route path="/" exact element={<Dashboard />} />                        
                        <Route path="/login" element={<LogIn image={image} />} />
                        <Route path="/new" element={<CreatePoll />} />
                        <Route path="/leaderboard" element={<Leaderboard />} />
                        <Route path="/questions/:pid" element={<PollPage image={image}/>} />
                      </Routes>
                    </Fragment>   
                  )           
                  : <LogIn image={image} />                  
              }            
            </div>
        )}
    </Fragment>
  );
};

const mapStateToProps = ({  authedUser, loading, users }, props) => {  
  // const keys = Object.keys(users);
  const { location } = props.router;
  return {
    loading: loading,
    authedUser,
    avatarURL: users[authedUser]?.avatarURL,
    location,
  }
    
    
};
  

export default withRouter(connect(mapStateToProps)(App));
