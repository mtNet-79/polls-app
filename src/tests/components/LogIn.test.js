import {
  waitForElementToBeRemoved,
  screen,
  waitFor,
} from "@testing-library/react";
import LogIn from "../../Components/LogIn";
import App from "../../Components/App";
import { renderWithProviders } from "./test-utils";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import "@testing-library/jest-dom/extend-expect";

// beforeEach(() => {
//     initializeCityDatabase();
//   });
  
//   afterEach(() => {
//     clearCityDatabase();
//   });

beforeEach(() => {
  delete window.location;
  // @ts-ignore
  window.location = new URL('http://localhost/');
});

//unit tests
test("all elements are rendered/ snapshot", () => {
  const {container} = renderWithProviders(<LogIn />);

  //the avatar holder is on the page
  const image = screen.getByRole("img");
  expect(image).toBeInTheDocument();

  // user input field present
  const userInpuEl = screen.getByPlaceholderText(/user name/i);
  expect(userInpuEl).toBeInTheDocument();

  //password present
  const passwaordInputEl = screen.getByPlaceholderText(/password/i);
  expect(passwaordInputEl).toBeInTheDocument();

  //submit button present
  const submitBtn = screen.getByRole("button", { name: /log in/i });
  expect(submitBtn).toBeInTheDocument();
  expect(submitBtn).toBeDisabled();

  // input count
  const inputsLogInPAge = screen.getAllByRole("textbox");
  expect(inputsLogInPAge.length).toBe(2);

  //add user link is present
  const addNewUserBtn = screen.getByRole("link", { name: /add new user/i });
  expect(addNewUserBtn).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test("log in button", async () => {

  const user = userEvent.setup();
  renderWithProviders(<LogIn />);

//  console.log("store at login 62: ", store)

  //in doc but disabled
  const submitBtn = screen.getByRole("button", { name: /log in/i });
  expect(submitBtn).toBeInTheDocument();
  expect(submitBtn).toBeDisabled();

  //enter text in user name input field
  const userInput = screen.getByRole("textbox", { name: /user/i });
  await userEvent.type(userInput, "mthornton");
  expect(userInput.value).toBe("mthornton");

  //button is still disabled
  expect(submitBtn).toBeDisabled();

  //enter password text to field - 5 characters 
  const passwordInput = screen.getByRole("textbox", { name: /password/i });
  await user.type(passwordInput, "pass1");
  expect(passwordInput.value.length).toBe(5);
  //disabled
  expect(submitBtn).toBeDisabled();

  //required 7 characters for password
  await user.type(passwordInput, "2");
  expect(passwordInput.value.length).toBe(6);

  //enabled
  expect(submitBtn).not.toBeDisabled();

});
//integration tests
test(`log in takes user to dashboard`, async () => {
  const user = userEvent.setup();

  const { container } = renderWithProviders(<App />, { route: "/login" });

  // screen.debug(console.log("store at login 98: ", store.getState()))

  await waitFor(
    () =>
      expect(
        screen.getByRole("button", { name: /log in/i })
      ).toBeInTheDocument(),
    {
      timeout: 2500,
    }
  );

  // screen.debug(console.log("store at login 110: ", store.getState()))

  await user.type(screen.getByLabelText(/user/i), "mthornton");
  const userInput = screen.getByRole("textbox", { name: /user/i });
  expect(userInput.value).toBe("mthornton");

  await user.type(screen.getByLabelText(/password/i), "elephants");

  user.click(screen.getByRole("button", { name: /log in/i }));
  await waitForElementToBeRemoved(() =>
    screen.getByRole("button", { name: /log in/i })
  );

  expect(screen.getByRole('heading', {name: /new polls/i})).toBeInTheDocument();
  //has naviagation
  const navigation = container.getElementsByClassName("nav");
  expect(navigation.length).toBe(1);
});
