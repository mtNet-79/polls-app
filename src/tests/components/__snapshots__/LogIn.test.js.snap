// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all elements are rendered/ snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <img
      alt="description"
      class="avatar"
    />
    <div
      class="add-new"
    >
      <form
        class="add-new"
      >
        <label
          class="form-label center"
          for="userName"
        >
          User
        </label>
        <input
          id="userName"
          name="userName"
          placeholder="User Name"
          type="text"
        />
        <label
          class="form-label center"
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          name="password"
          placeholder="Password"
          type="text"
        />
        <button
          class="btn log-in-button"
          disabled=""
          type="submit"
        >
          Log In
        </button>
      </form>
      <a
        class="btn add-new-btn"
        href="/add-user"
        value="Add New User"
      >
        Add New User
      </a>
    </div>
  </div>
</div>
`;
